src/logic/synchronization.ts

```diff
@@ -105,57 +124,70 @@ async function importMetadata(
     };
 }
 
-export async function startSynchronization(
+export async function* startSynchronization(
     d2: D2,
     builder: SynchronizationBuilder
-): Promise<SyncReport> {
+): AsyncIterableIterator<SynchronizationState> {
     const { targetInstances: targetInstanceIds, metadata } = builder;
 
     // Phase 1: Export and package metadata from origin instance
-    console.debug("Begin Export and package metadata from origin instance");
+    console.debug("Start synchronization process");
+    yield { message: i18n.t("Fetching metadata from origin instance") };
     const exportPromises = _.keys(metadata)
-        .map(type => {
-            const myClass = d2ModelFactory(d2, type);
-            return {
-                type,
-                ids: metadata[type],
-                excludeRules: myClass.getExcludeRules(),
-                includeRules: myClass.getIncludeRules(),
-            };
-        })
-        .map(newBuilder => exportMetadata(d2, newBuilder));
+        .map(
+            (type: string): ExportBuilder => {
+                const myClass = d2ModelFactory(d2, type);
+                return {
+                    type,
+                    ids: metadata[type],
+                    excludeRules: myClass.getExcludeRules(),
+                    includeRules: myClass.getIncludeRules(),
+                };
+            }
+        )
+        .map(
+            (newBuilder: ExportBuilder): Promise<MetadataPackage> => exportMetadata(d2, newBuilder)
+        );
     const exportResults: MetadataPackage[] = await Promise.all(exportPromises);
     const metadataPackage = _.deepMerge({}, ...exportResults);
     console.debug("Metadata package from origin instance done", metadataPackage);
 
     // Phase 2: Create parent synchronization task
+    yield { message: i18n.t("Retrieving information from remote instances") };
     const targetInstances: Instance[] = await Promise.all(
-        targetInstanceIds.map(id => Instance.get(d2, id))
+        targetInstanceIds.map((id: string): Promise<Instance> => Instance.get(d2, id))
```

Comment:
[minor] Do we need those explicit types? In a lambda, typically you onit the types if TS knows the types. There are some other instances of this in the code.

Reply:
We had places with them, and others without. I just added them everywhere. No strong opinion about leaving or keeping them.